### PR TITLE
Gpa 89 il2 cpp compiler

### PR DESCRIPTION
**Pull request by emeline-berenguier**
### Description

Add an analyzer for IL2CPP Compiler Configuration mode

### Changes made

_ Add two descriptors in PlayerSettingsAnalyzer : one to check Debug mode enabled, the other one to check Master mode enabled

### Notes

-

### Checklist

Before review:

- [x] Changelog entry added.
- [x] Tests added/changed, if applicable.
- [ ] Docs for new/changed API's.

### CODE REVIEW

File: Documentation~/Diagnostics.md [hunk 1]
```diff
@@ -53,6 +53,8 @@ This is a full list of all builtin settings diagnostics:
 | PAS1001 | Lit Shader Mode Forward and Deferred        | HDRP      | Any                      |
 | PAS1002 | Camera Lit Shader Mode Forward and Deferred | HDRP      | Any                      |
 | PAS1003 | Fog Mode Stripping						    | Graphics  | Any                      |
+| PAS1004 | IL2CPP Compiler Master						| Player    | Any                      |
+| PAS1004 | IL2CPP Compiler Debug						| Player    | Any                      |
```
Comment (mtrive) [inline Documentation~/Diagnostics.md:57]: should be _PAS1005_

File: Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs [hunk 2]
```diff
@@ -103,5 +144,16 @@ internal static void FixSpeakerMode()
 
             AudioSettings.Reset(audioConfiguration);
         }
+
+        internal static bool CheckIL2CPPCompilerConfiguration(Il2CppCompilerConfiguration compilerConfiguration)
+        {
+            return PlayerSettings.GetIl2CppCompilerConfiguration(EditorUserBuildSettings.selectedBuildTargetGroup) ==
```
Comment (mtrive) [inline Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs:150]: before this, I would return false if Mono scripting backend is in use

File: Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs [hunk 3]
```diff
@@ -44,11 +44,42 @@ class PlayerSettingsAnalyzer : ISettingsModuleAnalyzer
             })
         };
 
+        static readonly Descriptor k_IL2CPPCompilerConfigurationMasterDescriptor = new Descriptor(
+            "PAS1004",
```
Comment (mtrive) [inline Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs:48]: I changed how we declare IDs. Let's use [this ](https://github.com/Unity-Technologies/ProjectAuditor/blob/3a8465ee4b4ca454932a354daae62a87eabff985/Editor/SettingsAnalysis/GraphicsApiAnalyzer.cs#L17) pattern. By doing that, we can replace all hardcode string IDs accordingly.

File: Tests/Editor/SettingsAnalysisTests.cs [hunk 4]
```diff
@@ -342,5 +342,46 @@ public void SettingsAnalysis_FogStripping_IsNotReported(FogStripping fogModeStri
...
+                    issues = Analyze(IssueCategory.ProjectSetting, i => i.descriptor.id.Equals("PAS1004"));
+                    break;
+                case  Il2CppCompilerConfiguration.Debug:
+                    issues = Analyze(IssueCategory.ProjectSetting, i => i.descriptor.id.Equals("PAS1005"));
+                    break;
+            }
+
+            var playerSettingIssue = issues.FirstOrDefault();
+
+            Assert.NotNull(playerSettingIssue);
```
Comment (mtrive) [inline Tests/Editor/SettingsAnalysisTests.cs:369]: let's also check only 1 issue is reported.

File: Tests/Editor/SettingsAnalysisTests.cs [hunk 5]
```diff
@@ -342,5 +342,46 @@ public void SettingsAnalysis_FogStripping_IsNotReported(FogStripping fogModeStri
...
+
+            PlayerSettings.SetIl2CppCompilerConfiguration(EditorUserBuildSettings.selectedBuildTargetGroup, compilerConfiguration);
+        }
+
```
Comment (mtrive) [inline Tests/Editor/SettingsAnalysisTests.cs:373]: could you please add two additional tests:
- _SettingsAnalysis_Il2CppCompilerConfigurationRelease_IsNotReported_
- _SettingsAnalysis_Il2CppCompilerConfigurationMaster_ScriptingBackendMono_IsNotReported_

File: Tests/Editor/SettingsAnalysisTests.cs [hunk 6]
```diff
@@ -342,5 +342,46 @@ public void SettingsAnalysis_FogStripping_IsNotReported(FogStripping fogModeStri
...
+        [Test]
+        [TestCase(Il2CppCompilerConfiguration.Debug)]
+        [TestCase(Il2CppCompilerConfiguration.Master)]
+        public void SettingsAnalysis_IL2CPP_Compiler_Configuration_IsReported(
+            Il2CppCompilerConfiguration il2CppCompilerConfiguration)
+        {
+            var compilerConfiguration = PlayerSettings.GetIl2CppCompilerConfiguration(EditorUserBuildSettings.selectedBuildTargetGroup);
+            PlayerSettings.SetIl2CppCompilerConfiguration(EditorUserBuildSettings.selectedBuildTargetGroup, il2CppCompilerConfiguration);
+
```
Comment (mtrive) [inline Tests/Editor/SettingsAnalysisTests.cs:354]: I think we should also set/restore ScriptingBackend to Il2Cpp

File: Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs [hunk 7]
```diff
@@ -44,11 +44,42 @@ class PlayerSettingsAnalyzer : ISettingsModuleAnalyzer
...
+            new[] { Area.CPU },
+            "<b>C++ Compiler Configuration</b> is set to <b>Debug</b>. The performances will be suboptimal. Keep this mode only for debugging only.",
+            "To have optimal performances, change <b>Project Settings ➔ Configuration ➔ C++ Compiler Configuration</b> to <b>Release</b>.")
+        {
+            fixer = (issue =>
+            {
+                SetIL2CPPConfigurationToRelease();
+            }),
+
+            messageFormat = "Player : C++ Compiler Configuration is set to 'Debug'. The performances will be suboptimal."
```
Comment (mtrive) [inline Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs:73]: I would omit the " The performances will be suboptimal." as this is already in the description.

File: Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs [hunk 8]
```diff
@@ -44,11 +44,42 @@ class PlayerSettingsAnalyzer : ISettingsModuleAnalyzer
...
+            "Player: IL2CPP Compiler Configuration",
+            new[] { Area.BuildTime },
+            "<b>C++ Compiler Configuration</b> is set to <b>Master</b>. The build time will be longer. Keep this mode only for shipping builds.",
+            "To have optimal build time, change <b>Project Settings ➔ Configuration ➔ C++ Compiler Configuration</b> to <b>Release</b>.")
+        {
+            fixer = (issue => {
+                SetIL2CPPConfigurationToRelease();
+            }),
+
+            messageFormat = "Player : C++ Compiler Configuration is set to 'Master'. The build time will be longer."
```
Comment (mtrive) [inline Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs:58]: I would omit the " The build time will be longer." as this is already in the description.

File: Tests/Editor/SettingsAnalysisTests.cs [hunk 9]
```diff
@@ -342,5 +342,46 @@ public void SettingsAnalysis_FogStripping_IsNotReported(FogStripping fogModeStri
...
+            ProjectIssue[] issues = null;
+
+            switch (il2CppCompilerConfiguration)
```
Comment (mtrive) [inline Tests/Editor/SettingsAnalysisTests.cs:357]: How about using a more compact and efficient version like:
```
var id = il2CppCompilerConfiguration == Il2CppCompilerConfiguration.Master 
    ? "PAS1004"
    : "PAS1005";

var issues = Analyze(IssueCategory.ProjectSetting, i => i.descriptor.id.Equals(id));
```

**Review by mtrive — CHANGES_REQUESTED** [review]
I left several minor comments but overall looks good!

File: Tests/Editor/SettingsAnalysisTests.cs (hunk 9 above)
Comment (emeline-berenguier) [reply Tests/Editor/SettingsAnalysisTests.cs:357]: ChatGPT ;)

File: Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs [hunk 10]
```diff
@@ -103,5 +150,23 @@ internal static void FixSpeakerMode()
 
             AudioSettings.Reset(audioConfiguration);
         }
+
+        internal static bool CheckIL2CPPCompilerConfiguration(Il2CppCompilerConfiguration compilerConfiguration)
+        {
+            var buildTargetGroup = EditorUserBuildSettings.selectedBuildTargetGroup;
```
Comment (mtrive) [inline Editor/SettingsAnalysis/PlayerSettingsAnalyzer.cs:156]: the selected build target is not necessarily the platform we are analyzing. You can get that via `projectAuditorParams`

**Review by mtrive — APPROVED** [review]
Approved subject to:
- fixing `CheckIL2CPPCompilerConfiguration `to check analysis platform
- simplify `messageFormat `so it does not include redundant info

Great work!

**Pull request merged**